scripts/fuzz_shell.js

```diff
@@ -423,16 +408,74 @@ function hashCombine(seed, value) {
       continue;
     }
 
+    // A task is a name + a function to call. For an export, the function is
+    // simply a call of the export.
+    tasks.push({ name: name, func: /* async */ () => callFunc(value) });
+  }
+
+  // Reverse the array, so the first task is at the end, for efficient
+  // popping in the common case.
+  tasks.reverse();
+
+  // Execute tasks while they remain.
+  while (tasks.length) {
+    let task;
+    if (ordering === undefined) {
+      // Use the natural order.
+      task = tasks.pop();
+    } else {
+      // Pick a random task.
+      ordering = hashCombine(ordering, tasks.length);
+      let i = ordering % tasks.length;
+      task = tasks.splice(i, 1)[0];
+    }
+
+    // Execute the task.
+    console.log('[fuzz-exec] calling ' + task.name);
+    let result;
     try {
-      console.log('[fuzz-exec] calling ' + name);
-      // TODO: Based on |ordering|, do not always await, leaving a promise
-      //       for later, so we interleave stacks.
-      var result = /* await */ callFunc(value);
-      if (typeof result !== 'undefined') {
-        console.log('[fuzz-exec] note result: ' + name + ' => ' + printed(result));
-      }
+      result = task.func();
     } catch (e) {
       console.log('exception thrown: ' + e);
+      continue;
+    }
+
+    if (JSPI) {
+      // When we are changing up the order, in JSPI we can also leave some
+      // promises unresolved until later, which lets us interleave them. Note we
+      // never defer a task more than once (which would be pointless), and we
```

Comment:
Deferring a task more than once would be interesting if it were deferred via chaining to another promise. And it would be even more interesting if it were "chained" to another promise by being returned from a future JSPI import call.

Reply:
Ah, interesting, what do you mean by "chaining" here? Is there an API for that?

Reply:
`var newPromise = oldPromise.then((x) => { return x; });` or similar ways of wrapping the contents of the old promise in a new promise that will only be resolved after the old promise.

Reply:
Do you think that would test anything on the wasm side? I'd guess that just defers any wasm work further? I don't really know the VM side though.

As for chaining to a future JSPI call, that does happen already, if there is this in the wasm:
```
sleep();
sleep();
sleep();
```
Each one sleeps, then we continue, then we sleep again etc., which I think achieves the same?

Reply:
Yeah, maybe chaining in pure JS wouldn't add much. I think chaining through future JSPI calls might still be interesting because it would mix up the temporal order of tasks. The common case is that a task will not be resumed until a subsequently started task is completed, as in the sequence of sleeps above, but it is also interesting to test tasks that will not be resumed until prior tasks are completed.

Reply:
Added a todo.